modules/diskq/tests/test_diskq_truncate.c

```diff
@@ -380,6 +388,134 @@ Test(diskq_truncate, test_diskq_truncate_size_ratio_1)
+
+  // 4. send and ack all messages
+  send_some_messages(q, unprocessed_messages_in_buffer);
+  log_queue_ack_backlog(q, unprocessed_messages_in_buffer);
+  cr_assert(qdisk_get_length(qdisk) == 0, "qdisk len should be 0");
+  unprocessed_messages_in_buffer = 0;
+  unprocessed_messages_in_buffer = 0;
+
+  // 5. feed to full
+  feed_some_messages(q, just_under_max_size_message_number);
+  cr_assert_eq(log_queue_get_length(q), just_under_max_size_message_number,
+               "Not all messages have been queued!");
+  unprocessed_messages_in_buffer += just_under_max_size_message_number;
+
+  // 6. process some messages, so write_head can wrap
+  send_some_messages(q, some_messages_to_let_write_head_wrap);
+  log_queue_ack_backlog(q, some_messages_to_let_write_head_wrap);
+  unprocessed_messages_in_buffer -= some_messages_to_let_write_head_wrap;
+
+  // 7. feed 2 to wrap write_head and to add one message to the beginning
+  feed_some_messages(q, 1);
+  cr_assert(qdisk_get_writer_head(qdisk) == QDISK_RESERVED_SPACE, "write_head should have wrapped");
+  feed_some_messages(q, 1);
+  unprocessed_messages_in_buffer += 2;
+
+  // 8. process all messages
+  send_some_messages(q, unprocessed_messages_in_buffer);
+  log_queue_ack_backlog(q, unprocessed_messages_in_buffer);
+  cr_assert(qdisk_get_length(qdisk) == 0, "qdisk len should be 0");
+  cr_assert(qdisk_get_writer_head(qdisk) == qdisk_get_reader_head(qdisk), "read_head should be at write_head's position");
+  unprocessed_messages_in_buffer = 0;
+
+  _assert_diskq_actual_file_size_with_stored(q);
+
+  unlink(filename->str);
+  g_string_free(filename, TRUE);
+
+  _save_diskqueue(q);
+}
+
+Test(diskq_truncate, test_non_reliable_diskq_restart_with_truncation_disabled)
+{
+  const gint qout_size = 128;
+  const gint just_under_max_size_message_number = 8239 + qout_size; // measured for disk_buf_size TEST_DISKQ_SIZE
+  const gint some_messages_to_let_write_head_wrap = 500;
+
+  GString *filename = g_string_new("test_dq_non_reliable_restart.rqf");
+
+  DiskQueueOptions options;
+  LogQueue *q = _create_non_reliable_diskqueue(filename->str, &options, qout_size, 1);
+  cr_assert_eq(log_queue_get_length(q), 0, "No messages should be in a newly created disk-queue file!");
+
+  QDisk *qdisk = ((LogQueueDisk *)q)->qdisk;
+
+  // 1. feed to full
+  feed_some_messages(q, just_under_max_size_message_number);
+  gint unprocessed_messages_in_buffer = just_under_max_size_message_number;
+
+  // 2. process some messages, so write_head can wrap
+  send_some_messages(q, some_messages_to_let_write_head_wrap);
+  log_queue_ack_backlog(q, some_messages_to_let_write_head_wrap);
+  unprocessed_messages_in_buffer -= some_messages_to_let_write_head_wrap;
+
+  // 3. feed additional messages, so write_head wraps
+  feed_some_messages(q, 300);
+  unprocessed_messages_in_buffer += 300;
+  cr_assert(qdisk_get_writer_head(qdisk) < qdisk_get_reader_head(qdisk), "write_head should have wrapped");
+
+  // Restart
```

Comment:
I think it is a really good "side effect" of the new truncation feature that we started to test on the number of messages instead of pure file size. (They were technically equal before, but it was semantically incorrect.)

However the current implementation (and I think a unit test can rely on that detail) depends on the fact that we truncate away the non reliable queue's (qout, qbacklog, qoverflow), even if the truncation is completely disabled.

As a result: my proposal is that we should check on the actual file size after load, prior starting to send out any messages.

Reply:
IMHO the current implementation that truncates memory-buffers away should be fixed eventually.
dqtool open the disk queue in read-only mode, so there were bugs there too.